onc-healthit/inferno-program
# FI-1176: Ensure code exists when doing code search with system.

# Summary
Our searches are set up to choose which values to populate search criteria with based on data that has already been seen.  Since it is technically valid to not populate many values in the resource, including values which may be searched by, we have to ensure that we do not pick values that are null / filled with DARs.

We are doing this properly for most searches.  However, it turns out we are not doing this properly when we re-do a "code" search with the equivalent "system|code" search.  So for the first search (just by code), we locate and use a valid code, but then when we redo the search with system, our code picks up a different resource with a code not filled in, if we are unlucky.

This only occurs if the first resource returned has no code to search by, so you have to be a little unlucky for this to happen, and which is why we hadn't seen this before.  This is what the query looks like:

![Screen Shot 2021-03-08 at 9 54 32 PM](https://user-images.githubusercontent.com/412901/110412366-156fea00-805a-11eb-87f8-3612d23f263b.png)

Note the invalid search in the second request.

This addresses #248 

## New behavior

The search by system+code now will pick up the same code as the "just by code" search.  Note the valid second request now:

![Screen Shot 2021-03-08 at 9 57 28 PM](https://user-images.githubusercontent.com/412901/110412462-3fc1a780-805a-11eb-8ef0-183124851199.png)

## Code changes

The generated was updated slightly to include the same guard for picking codes that aren't filled in.

## Testing guidance

Unfortunately updating our generated unit tests to try this out seems very challenging.  I tested it manually by:

1) Running the reference server locally
2) Running Inferno's Lab Result Observation test against the reference server (I updated the onc_program_module.yml file to only include a couple of the single patient api tests to speed things up)
3) Noting which resource was returned first in USCLRO-02.
4) Since there was only one resource, I duplicated it with Postman
5) Run again, check to see which resource was returned first
6) Update the resource that was returned first to only have a "text" for "Observation.code"
7) Run again, check to make sure that the test passes
8) Revert to the old version, check to make sure that the test fails
